### PR TITLE
Add Modal property overlay color

### PR DESCRIPTION
**Pull request by 7thPorter**
Full disclosure: This feature was originally suggested [here](https://github.com/facebook/react-native/issues/18398) in 2018, but no pull request was made and the issue was closed. I can’t take credit for the idea originally, but since the original poster’s GitHub account no longer exists for me to ask permission, and the issue in question persists through today, I decided to take a shot at a fix. For reference, another attempt was made to fix this same issue [here](https://github.com/facebook/react-native/pull/24949), but the request was closed by the requester, rather than merged.

## Summary

Currently, when a developer sets the _transparency_ prop of a Modal component to false, a white page is rendered “behind” the modal. This effect is hardcoded. This small feature allows developers to utilize a new prop, called **overlayColor**, to change this background color to suit their needs. The value of **overlayColor** will be a string, and may include anything that could be entered as a valid _backgroundColor_ style. 

In order to facilitate possible use of transparency values in the **overlayColor** string (i.e. rgba values), the _presentationStyle_ prop logic has been modified to make “overFullScreen” the default value of that prop when **overlayColor** is utilized. Without making this change to the _presentationStyle_ prop, color values with transparency cannot be rendered properly on iOS.

## Changelog

[General] [Added] - Add overlayColor property to Modal component

## Test Plan

**npm test result**: When testing, I first ran the test on the main branch as a control group and received 196 failed tests of 2845 total. Not sure what causes those failures, but after implementing changes on the modal-overlayColor branch and running npm test again, the number of failed tests did not increase.

**npm run flow result**: The flow test showed that I hadn’t properly considered all the pieces of the Modal component. 

First, and in particular, checking for the value of **overlayColor** as a string returned a “sketchy-null-string” error. Investigation led me to the Text component, where a similarly-functioning feature, the _selectionColor_ prop, imports processColor from the StyleSheet component, and passes the input string through that function. Doing so in my code eliminated the sketchy-null-string error, but reveals a “sketchy-null-number” error. If possible, I’d appreciate some guidance on why this is happening, and how to resolve it, since I had never used flow before attempting these tests. In the end, I settled on a solution that works in my test code, but still gives me a couple "sketchy-null-string" error. I don’t feel like this is necessarily the most optimized solution, so if there’s a better way to approach this, I’d be eager to hear it.

Second, it showed that I needed to edit the RCTModalHostViewNativeComponent file to reflect changes made in Modal.js. Again, I relied on the TextNativeComponent file for usage of the _selectionColor_ prop as a guide.

#### Default Behavior:
React Native Modal - defaultBehavior
#### With **overlayColor** Enabled:
React Native Modal - overlayColor

_**Note 1**_: Allowing the use of transparency values in the **overlayColor** prop when the _transparent_ prop is set to false may be a bit misleading. There's probably a better solution available to remove any possible confusion, but I've taken to considering _transparent_ to mean "true transparency", whereas the rgba value represented in the screenshot is only a partial transparency.

_**Note 2**_: Because of the behavior of the Modal component, setting the _transparent_ prop to false results in an unaesthetic animation in the background color if the _animationType_ property is set to "slide". This doesn't fall within the scope of this pull request, but it's probably worth noting for developers so they can expect that behavior.

### CODE REVIEW

**Comment by charlesbdudley** [comment]
Thanks for this contribution! Could you take a look at the eslint and flowtype errors?

**Comment by 7thPorter** [comment]
Hi @charlesbdudley this is what I was looking for guidance with, actually. Here are the results of running the flow test:

```
Error ---------------------------------------------------------------------------------- Libraries/Modal/Modal.js:243:13

Sketchy null check on string [1] which is potentially an empty string. Perhaps you meant to check for null or
undefined [2]? [sketchy-null-string]

 [2][1]  96|   overlayColor?: ?string,
           :
        240|       backgroundColor:
        241|         this.props.transparent === true
        242|           ? 'transparent'
        243|           : this.props.overlayColor
        244|           ? this.props.overlayColor
        245|           : 'white',
        246|     };
```
I don't have any experience with flow outside of running these tests, so I'm not quite sure what to do about this sketchy-null-string error. I tried my best to eliminate the errors by performing null checks (as on lines 234-237 of the Modal.js file in the PR), but when I worked this last bit of code to finally resolve the null-string error, it made the overlayColor prop non-functional, so I was left with that last error. Is there a good way to address this?

I'm also not sure I installed all of my lint packages correctly, since I am getting 1030 errors when I run "npm run lint", and most, if not all of them, have nothing to do with the changes I made. Is this something commonly run into when contributors are testing their code? I feel very silly not getting this right, especially for some relatively small changes to the codebase 😓

File: Libraries/Modal/Modal.js [hunk 1]
```diff
@@ -223,17 +231,26 @@ class Modal extends React.Component<Props> {
...
+      this.props.overlayColor == null
+        ? null
+        : processColor(this.props.overlayColor);
+
     const containerStyles = {
       backgroundColor:
-        this.props.transparent === true ? 'transparent' : 'white',
+        this.props.transparent === true
+          ? 'transparent'
+          : this.props.overlayColor
```
Comment (levibuzolic) [inline Libraries/Modal/Modal.js:243]: Should sort out the sketchy null check.

```suggestion
          : this.props.overlayColor != null
```

**Comment by alexgshaw** [comment]
Is this not happening anymore? Would love this feature.

**Pull request closed without merging**